### DISCUSSION TITLE
Frictionless Python Framework Roadmap: Limits of the Package

### DISCUSSION THREAD

**Discussion by pdelboca** [frictionlessdata/frictionless-py#1723] [General]
We just had an excellent conversation about the future and the roadmap of the Frictionless Python Framework. We talked about:
- the current status, 
- the work needed, 
- support for v2 of Datapackage, 
- a common API with other framewors (like the R package)
- and more! 

We agreed that an excellent next step will be to define the limits of the Python Framework both in terms of what we want it to do and what we do not want it to do anymore (as a preparation to define how the API should look like in the future) The idea is to come up with an agreed list of items that will guide future actions.

The benefits of having a clear set of limits are:
 - Make it more maintainable by reducing and limiting the scope.
 - Create consensus among the community on the future.
 - Align expectations.
 - Empower maintainers to take action and implement changes.

I'll kickoff with some points based on the discussion we had, feel free to adapt/add/remove items.

### What `frictionless-py` should do:
- xxx
- xxx

### What `frictionless-py` should not do:
- It should not be a swiss knife for all the task related to data analysis. (Let's expand this into more explicit items. What things in the code should no longer exist?)
- It should not implement support for data portals.

**Comment by pierrecamilleri** [comment]
Thanks a lot Patricio for continuing this conversation. 

Before giving my opinion, I'd like to make it clear that I'm probably biased by our use case and have limited knowledge of other use cases, so I'm obviously (and as always!) open to any contradiction. 

That said, I come from the school where a tool has to do one thing and do it well, and for frictionless, I think that what brings the most value is data validation.

So for me, and as a first take:

**What `frictionless-py` should do:**

- Validate table data, accepting numerous data sources, with the Table Schema and Data Package specifications.
    - I think providing explicit errors in case of malformed csv, or malformed json, is valuable (or any other non-binary format). Ideally, I'd prefer this to be handled by a dependency.  
    - Areas for improvement in this area (apart from adapting to DP v2) include [performance](https://github.com/frictionlessdata/frictionless-py/issues/1721), documentation and ease of use (some options for Excel, for example, are [fairly cryptic](https://framework.frictionlessdata.io/docs/formats/excel.html#reference-formats.excelcontrol.stringified)), extensibility (writing new checks today requires [a certain familiarity with frictionless-py code](https://github.com/frictionlessdata/frictionless-py/issues/1612)), native support for [i18n](https://github.com/frictionlessdata/frictionless-py/issues/544) and this with no compromise to reliability.

**What `frictionless-py` should not do:**

- Data transformation. Many tools probably do this better. 
- Data portals - as mentioned in our conversation, some of this functionality has been transferred to other tools (Zenodo was mentioned).

I have mixed feelings about this one : 

- Guesswork. Guess file format, infer column formats etc. While guessing these may seem attractive as a way of hiding complexity from the user, it's the source of many bugs (to name but a few, [this one](https://github.com/frictionlessdata/frictionless-py/issues/1646), [this one](https://github.com/frictionlessdata/frictionless-py/issues/1607) and [this one](https://github.com/frictionlessdata/frictionless-py/issues/1714)), it introduces real complexity into the code, it complicates documentation and typing and I don't think it's worth it: the user usually knows the format of the data he wants to validate. However, I acknowledge that, especially for the command line client, is is nice to not have to enter 4 options to be able to perform a validation - but to wrap it up, I would prefer the validation to fail if there is some uncertainty rather than being unreliable for edge case.

**Reply by peterdesmet** [reply]
I agree. I (only) use frictionless-py for its full-feature **validation**. It's such a massive feature that I didn't replicate it in R. @mpadge and I discussed an idea once that it would be nice to have a fast Data Package validation library (in C??) that **could then be used by both frictionless-py and frictionless-r (and frictionless-java, etc.)**.

I think frictionless-py and frictionless-r should do:

- Input/output
- Human readable print
- Manipulation: adding/removing files and data frames as resources + adding/removing metadata properties
- Validation (using a common library that is not a burden to install) with human/machine readable error report (e.g. a data frame with the errors)

Their APIs could be discussed together, but I don't think they should necessarily be the same: they should make sense in their relative context. Frictionless-r for example is currently designed to get out of the way quickly, returning your data as a data frame that you can use in so many other R libraries. It was one of the things I was missing in the previous [datapackage-r](https://docs.ropensci.org/frictionless/#frictionless-vs-datapackager) package.

**Reply by khughitt** [reply]
> Their APIs could be discussed together, but I don't think they should necessarily be the same.

Agreed. Trying to force the APIs to be identical might make sense for something with a very narrow scope (e.g. [Arrow](https://github.com/apache/arrow)), but isn't likely to work for every project.

How about trying to coordinate on a "common core" API for the elements you listed above?

> Frictionless-r for example is currently designed to get out of the way quickly, returning your data as a data frame that you can use in so many other R libraries

This is also mostly how I think about things / what I'm most interested in as a _user_.

I wonder if it would be helpful to think about two (or more) different types of users:

1. people using `frictionless-py` / `frictionless-r` to interact with data effortlessly
2. people using `frictionless-py` / `frictionless-r` as part of larger systems for data ingestion, validation, etc.?

**Reply by peterdesmet** [reply]
> How about trying to coordinate on a "common core" API for the elements you listed above?

Yeah, definitely open for that.

The different types of users aren't necessarily useful to me, we should cater to both in the design of the API. It is why frictionless-r functions can be piped (`|>`) together:

```R
create_package() |>
  remove_resource("wrong")
  add_resource("data-1", df) |>
  add_resource("data-2", df) |>
  write_package()
```

**Comment by khughitt** [comment]
Thanks for staying on top of things and starting the conversation, Patricio!

I strongly agree with the "_do one thing and do it well_" approach..

As we discussed in the meeting yesterday, if we decide there are _other_ things we would like to do well, we can consider creating separate modular/dedicated libraries aimed at them (long-term ~> frictionless "_universe_"?)

Would it be useful to generalize this conversation, and pose the same questions for [frictionless-r](https://github.com/frictionlessdata/frictionless-r) now? (@peterdesmet ?)

Once we agree on the high-level points / find as much common ground as is sensible, we can then separately discuss what we want to do for the current language implementations.

At the end of this discussion, let's compile a list of action items / separate discussions to have?

e.g. "_What things in the code should no longer exist?_"

### What `frictionless-py` should do:

1. validation
2. ~~i/o~~ ([https://github.com/frictionlessdata/dplib-py](dplib-py)'s job now?)

### What `frictionless-py` should _not_ do:

1. "_transform your data_"

- I agree that this should probably _not_ be handled by `frictionless-py`..
- this is a massive and open-ended task
- it will depend heavily on the data type
- other dedicated libraries may already exist
- if we decide take this on, a dedicated `frictionless-transforms-py` library would be a better place, imo

2. CLI **(update 12/19)**

After thinking about it some more, I would probably move CLI from "maybe" -> "should _not_".

It doesn't make sense to force every language binding to implement a CLI interface.

Also, separating the API & CLI would lead to more modular and simpler code bases for each.

### What `frictionless-py` _maybe_ should do? :shrug:

1. "_Describe your data_" ?

This is useful, but it's also very open-ended and might be better to split off?

~~2. CLI?~~

~~`click` and `typer` make it pretty straight-forward to create simple CLI's these days..~~

~~If the CLI just mirrors the `frictionless-py` API, as it appears to be doing now, then this is perhaps fine.~~

~~I could also imagine the CLI being a separate effort that builds on top of and interfaces with the APIs in `frictionless-py`, and perhaps other "core" libraries we may eventually decide to create.~~

~~I have no strong opinion on this for now; just raising awareness..~~

### Questions

- [ ] what are the things `frictionless-py` currently does _besides_ validation?
- [ ] how does [https://github.com/frictionlessdata/dplib-py](dplib-py) fit in to this?
  - what is _its_ intended scope?
  - how does it relate to `frictionless-py`?
  - would it make sense to create a similar library in R, `dplib-r`?
- [ ] where should functionality for interfacing with other related efforts such as [RO-Crates](https://www.researchobject.org/ro-crate/) reside?

## Possible next steps based on discussion so far:

1. **Performance**

- [ ] profile code / identify bottlenecks
- [ ] create benchmark suite (separate repo), and start to test different approaches to address bottlenecks identified

2. **Scope**

- [ ] create explicit lists of things `frictionless-py` / `frictionless-r` currently do _besides_ validation?

**Comment by ipimpat** [comment]
I’d like to contribute to the ongoing discussion by outlining a few areas where `frictionless-py` could improve, based on my experiences and observations.

### Suggested Improvements for `frictionless-py`

1. **Enhance Documentation**  
   The current documentation could be improved to make it more accessible and user-friendly, especially for newcomers. A great example to follow is the [Laravel PHP framework](https://laravel.com/docs/11.x), which sets a high standard for clarity, detail, and usability. Improving the documentation would lower the entry barrier and foster a larger, more active community around Frictionless.

2. **Support Data Transformation**  
   Frictionless should natively include tools for data transformation, which currently feels unintuitive and could benefit from significant improvements. A model similar to [Dagster](https://dagster.io/) or R's [targets](https://books.ropensci.org/targets/) library for defining pipelines, or first-class support for existing tools within the Python ecosystem, would be ideal. Data transformation is a critical step in data workflows and should be tightly integrated into the platform.

3. **Expand Filesystem Support**  
   Adding support for more filesystems, such as SMB network shares, would be a significant enhancement. SMB remains widely used in enterprise environments for file sharing and storage, making support for it essential. Currently, SMB support outside of Windows is inconsistent, creating challenges for users working across platforms, such as those switching between Windows and Linux (Linux implementations of SMB are riddled with bugs and limitations, especially in desktop environments, SMB shares are almost completely unusable and MacOS users likely face similar difficulties).

   To address these cross-platform challenges, I have explored integrating fsspec's [`universal_pathlib`](https://github.com/fsspec/universal_pathlib) with Frictionless via a Frictionless extension. This integration adds first-class support for SMB, allowing users to simply write:

   ```python
   package = Package("smb://host/share/mydataset/datapackage.yml")  # globbing also works
   resource = Resource("smb://host/share/mydata.csv")
   ```
    Technically, this approach could support a broad range of filesystems currently supported by fsspec's [universal_pathlib](https://github.com/fsspec/universal_pathlib?tab=readme-ov-file#currently-supported-filesystems-and-protocols), extending Frictionless' capabilities to handle diverse storage protocols.

4. **Improve Error Handling**  
   The current way Frictionless handles exceptions is problematic. Instead of preserving stack traces, it "eats" exceptions and only prints error messages, discarding critical debugging information. This makes developing custom steps unnecessarily difficult. Introducing a verbose debug flag or a mode that preserves and outputs full exception details would greatly enhance the developer experience. Better error handling would also facilitate debugging and make the framework more transparent.

5. **Broaden Compatibility with Databases and File Formats**  
   Frictionless should extend its capabilities to support a wider range of databases and file formats. Legacy formats like SPSS’s `.sav` and SAS’s `.sas7bdat` are still heavily relied upon in the social sciences. These formats align well with Data Package’s resource specifications, and enhancing the standard to include their metadata would enable seamless integration.

6. **Become Enterprise-Friendly**  
   Many enterprise data sources require authentication, which should be more effectively supported by Frictionless. Additionally, the current Control concept is under-documented, making it difficult for users to fully utilize this feature. Clarifying and expanding on Control’s capabilities would help bridge this gap.

7. **Offer a Stable and Extensible API**  
   Frictionless should provide a well-documented, stable API for building extensions. Including detailed examples and best practices would lower the barrier for contributors and foster innovation within the ecosystem.

8. **Expand Portal and Cataloging Capabilities**  
   The portal and cataloging functionality is a unique and valuable feature of Frictionless that deserves expansion. For example, a network share with a standardized folder structure could serve as a simple, lightweight, or "poor man's" catalog or portal.

   Network shares remain a common and essential resource, frequently provided to students and employees at research organizations as a standard service. For example, a PhD student could leverage their project’s network share as a straightforward internal catalog for datasets and metadata, enabling them to effectively manage their data throughout its entire lifecycle—from collection or generation to processing and transforming to eventual publication in open data catalogs on the internet.

   I’ve implemented a similar solution with my [`universal_pathlib`](https://github.com/fsspec/universal_pathlib)-Frictionless extension. This approach could create simple, "poor man's" catalogs based on network shares and similar setups, with metadata managed using Frictionless (e.g., a `catalog.yaml` file in the root).

9. **Default to YAML for Descriptor Files**  
   Frictionless should consider defaulting to YAML format for descriptor files instead of JSON. YAML aligns more closely with the human-readable and editable nature of Frictionless’ descriptors. Many tools in the Python ecosystem prefer YAML for configuration due to its simplicity and clarity.

10. **Support Environment Variables and `.env` Files**  
   Frictionless should support the use of environment variables and `.env` files to simplify configuration management, especially for sensitive settings like database credentials and HTTP authorization headers. This approach is widely used across modern development workflows and aligns well with the principles of secure and flexible configuration.

     Frictionless could then automatically load these settings, allowing them to be referenced directly in descriptors or configuration files. This would enhance usability and promote best practices in credential management.

**Comment by pierrecamilleri** [comment]
>  create explicit lists of things frictionless-py [...] currently do besides validation?

Sorry, better late than never :face_with_open_eyes_and_hand_over_mouth: 

Here is a quick clustering of the features I know of in frictionless-py 

### Data/metadata Validation

- Validate data `frictionless validate` : 
    - validates if data is well formatted
    - validates schemas, datapackage, resources descriptors
    - validates data against data package / table schema specs
    - (implicit Table Dialect inference, can read and validate a wide range of formats and from a wide range of sources)

### Facilitate data/metadata exploration

- `frictionless query` import data package in a sqlite database 

- `frictionless script` import data package as pandas dataframes 

- `frictionless extract` retrieves data from a source (can for isntance return the data as a csv)

- Explore data-package / table schema descriptors : 
    - `frictionless list` lists resources for a table schema / data package

- Dataset exploration `frictionless explore`: 
    - Interface with visidata
    - Not sure about the added value of this ? Wouldn't a Visidata user know how to open the data ? 

### Data transformation

- `frictionless convert` includes conversion features.  has really rough edges : could not get it to work on `csv` to `xlsx` conversion.

- Database indexing : 
    - Only for DBs, not sure what the use case is ? Maybe in relation to pipelines ?

- Data steps + `transform` python function can define a transformation pipeline. (Not sure about the use at the command line)

### Data publishing

-  `frictionless publish`  publish to data portals: 
    - CKAN, zenodo, Github

### Format inference

-   `frictionless describe` dataset format inference : 
    - Print columns and infer associated type.
    - `--json` option allows for getting a starting point for table schema and table dialect
